backend/src/auth_service/user_handles/mod.rs

```diff
@@ -0,0 +1,155 @@
+// SPDX-FileCopyrightText: 2025 Phoenix R&D GmbH <hello@phnx.im>
+//
+// SPDX-License-Identifier: AGPL-3.0-or-later
+use displaydoc::Display;
+use phnxtypes::{
+    credentials::keys::HandleVerifyingKey,
+    identifiers::{
+        USER_HANDLE_VALIDITY_PERIOD, UserHandle, UserHandleHash, UserHandleHashError,
+        UserHandleValidationError,
+    },
+    time::ExpirationData,
+};
+use thiserror::Error;
+use tonic::Status;
+use tracing::{error, warn};
+
+use super::AuthService;
+
+pub(crate) use persistence::UserHandleRecord;
+
+mod persistence;
+
+impl AuthService {
+    pub(crate) async fn as_create_handle(
+        &self,
+        verifying_key: HandleVerifyingKey,
+        handle_plaintext: String,
+        hash: UserHandleHash,
+    ) -> Result<(), CreateHandleError> {
+        let handle = UserHandle::new(handle_plaintext)?;
+
+        let local_hash = handle.hash()?;
+        if local_hash != hash {
+            return Err(CreateHandleError::HashMismatch);
+        }
+
+        let expiration_data = ExpirationData::new(USER_HANDLE_VALIDITY_PERIOD);
+
+        let record = UserHandleRecord {
+            user_handle_hash: hash,
+            verifying_key,
+            expiration_data,
+        };
+
+        if record.store(&self.db_pool).await? {
+            Ok(())
+        } else {
+            Err(CreateHandleError::UserHandleExists)
+        }
```

Comment:
Since we don't systematically clean expired handles, we should not fail when one still exists. We should reclaim it instead.